File: Sources/Application/ProfileQuotaSchedulePolicy.swift
import Foundation

enum ProfileQuotaQueryTrigger: Equatable, Sendable {
  case automatic
  case manual
}

struct ProfileQuotaSchedulePolicy {
  static let manualRefreshCooldown: TimeInterval = 60
  static let maximumAutomaticRetriesPerDay = 3

  private static let retryDelays: [TimeInterval] = [300, 900, 3_600]

  let calendar: Calendar

  init(calendar: Calendar = .current) {
    self.calendar = calendar
  }

  func dueDate(
    for subscription: TrackedSubscription,
    state: ProfileQuotaQueryState?,
    now: Date
  ) -> Date {
    guard state?.lastQueriedURLFingerprint == subscription.urlFingerprint else {
      return now
    }
    if let nextAttemptAt = state?.nextAttemptAt {
      return nextAttemptAt
    }
    guard let lastAttemptAt = state?.lastAttemptAt else {
      return now
    }
    return lastAttemptAt.addingTimeInterval(refreshInterval(for: subscription))
  }

  func successfulState(
    for subscription: TrackedSubscription,
    at date: Date,
    jitter: TimeInterval
  ) -> ProfileQuotaQueryState {
    ProfileQuotaQueryState(
      subscriptionID: subscription.id,
      lastAttemptAt: date,
      nextAttemptAt: date.addingTimeInterval(refreshInterval(for: subscription) + jitter),
      lastQueriedURLFingerprint: subscription.urlFingerprint,
      consecutiveFailures: 0,
      retryDayStart: nil,
      automaticRetryCount: 0
    )
  }

  func failedState(
    for subscription: TrackedSubscription,
    previous: ProfileQuotaQueryState?,
    trigger: ProfileQuotaQueryTrigger,
    at date: Date,
    jitter: TimeInterval
  ) -> ProfileQuotaQueryState {
    let dayStart = calendar.startOfDay(for: date)
    let isSameRetryDay = previous?.retryDayStart == dayStart
    let previousRetryCount = isSameRetryDay ? previous?.automaticRetryCount ?? 0 : 0
    let wasAutomaticRetry = trigger == .automatic && (previous?.consecutiveFailures ?? 0) > 0
    let retryCount = previousRetryCount + (wasAutomaticRetry ? 1 : 0)
    let failureCount = (previous?.consecutiveFailures ?? 0) + 1
    let hasReachedLimit = retryCount >= Self.maximumAutomaticRetriesPerDay
    let nextAttemptAt =
      hasReachedLimit
      ? date.addingTimeInterval(refreshInterval(for: subscription) + jitter)
      : date.addingTimeInterval(retryDelay(for: failureCount))

    return ProfileQuotaQueryState(
      subscriptionID: subscription.id,
      lastAttemptAt: date,
      nextAttemptAt: nextAttemptAt,
      lastQueriedURLFingerprint: subscription.urlFingerprint,
      consecutiveFailures: hasReachedLimit ? 0 : failureCount,
      retryDayStart: dayStart,
      automaticRetryCount: retryCount
    )
  }

  func manualRefreshAvailableAt(state: ProfileQuotaQueryState?) -> Date? {
    state?.lastAttemptAt?.addingTimeInterval(Self.manualRefreshCooldown)
  }

  private func refreshInterval(for subscription: TrackedSubscription) -> TimeInterval {
    TimeInterval((subscription.refreshIntervalMinutes ?? 360) * 60)
  }

  private func retryDelay(for failureCount: Int) -> TimeInterval {
    let index = min(max(failureCount - 1, 0), Self.retryDelays.count - 1)
    return Self.retryDelays[index]
  }
}

File: platform/windows/MihomoMeter.Windows.Core/Application/ProfileQuotaSchedulePolicy.cs
using MihomoMeter.Windows.Core.Domain;

namespace MihomoMeter.Windows.Core.Application;

public enum ProfileQuotaQueryTrigger
{
    Automatic,
    Manual,
}

public sealed class ProfileQuotaSchedulePolicy
{
    public static readonly TimeSpan ManualRefreshCooldown = TimeSpan.FromSeconds(60);
    public const int MaximumAutomaticRetriesPerDay = 3;
    private static readonly TimeSpan[] RetryDelays =
    {
        TimeSpan.FromMinutes(5),
        TimeSpan.FromMinutes(15),
        TimeSpan.FromHours(1),
    };

    public DateTimeOffset DueDate(
        TrackedSubscription subscription,
        ProfileQuotaQueryState? state,
        DateTimeOffset now)
    {
        if (!string.Equals(
                state?.LastQueriedUrlFingerprint,
                subscription.UrlFingerprint,
                StringComparison.Ordinal))
        {
            return now;
        }

        if (state?.NextAttemptAt is DateTimeOffset nextAttempt)
        {
            return nextAttempt;
        }

        return state?.LastAttemptAt is DateTimeOffset lastAttempt
            ? lastAttempt + RefreshInterval(subscription)
            : now;
    }

    public bool CanRefreshManually(
        ProfileQuotaQueryState? state,
        DateTimeOffset now)
    {
        if (state?.LastAttemptAt is not DateTimeOffset lastAttempt)
        {
            return true;
        }

        if (state.LastFailureCategory is "timeout" or "network")
        {
            return true;
        }

        return now >= lastAttempt + ManualRefreshCooldown;
    }

    public ProfileQuotaQueryState Success(
        TrackedSubscription subscription,
        DateTimeOffset now,
        TimeSpan jitter)
    {
        return new ProfileQuotaQueryState(
            subscription.Id,
            now,
            now + RefreshInterval(subscription) + jitter,
            subscription.UrlFingerprint,
            0,
            null,
            0,
            null);
    }

    public ProfileQuotaQueryState Failure(
        TrackedSubscription subscription,
        ProfileQuotaQueryState? previous,
        ProfileQuotaQueryTrigger trigger,
        DateTimeOffset now,
        TimeSpan jitter,
        string category)
    {
        var dayStart = new DateTimeOffset(now.Year, now.Month, now.Day, 0, 0, 0, now.Offset);
        var sameRetryDay = previous?.RetryDayStart == dayStart;
        var previousRetryCount = sameRetryDay ? previous?.AutomaticRetryCount ?? 0 : 0;
        var wasAutomaticRetry = trigger == ProfileQuotaQueryTrigger.Automatic
            && (previous?.ConsecutiveFailures ?? 0) > 0;
        var retryCount = previousRetryCount + (wasAutomaticRetry ? 1 : 0);
        var failureCount = (previous?.ConsecutiveFailures ?? 0) + 1;
        var reachedLimit = retryCount >= MaximumAutomaticRetriesPerDay;
        var delay = reachedLimit
            ? RefreshInterval(subscription) + jitter
            : RetryDelays[Math.Min(failureCount - 1, RetryDelays.Length - 1)];
        return new ProfileQuotaQueryState(
            subscription.Id,
            now,
            now + delay,
            subscription.UrlFingerprint,
            reachedLimit ? 0 : failureCount,
            dayStart,
            retryCount,
            category);
    }

    private static TimeSpan RefreshInterval(TrackedSubscription subscription)
    {
        return TimeSpan.FromMinutes(subscription.RefreshIntervalMinutes ?? 360);
    }
}
